File: scripts/components/create_organization_connection.py
#!/usr/bin/env python3
"""
create_organization_connection.py

PURPOSE: Create organization-managed connections via Prismatic GraphQL API using
         scoped config variables (per-customer connections managed by org).

USAGE: python create_organization_connection.py \\
    --component-key <key> \\
    --connection-key <key> \\
    --name <name> \\
    [--api-key <value>] \\
    [--base-url <value>] \\
    [--stable-key <key>] \\
    [--skip-test-connection]

PARAMETERS:
  --component-key        Component key (e.g., acme, tulyp)
  --connection-key       Connection key from component (e.g., acmeApiKey)
  --name                 Human-readable name for the connection
  --api-key              Optional API key value to set
  --base-url             Optional base URL value to set
  --stable-key           Optional stable key (defaults to {component-key}-api-key)
  --skip-test-connection Skip creating test customer config variable

HOW IT WORKS:
  Creates a two-tier connection structure:

  1. Scoped Config Variable (Template)
     - variableScope: "customer" (each customer gets their own instance)
     - managedBy: "org" (organization provides credentials, not customers)
     - Inputs have meta but NO values (values go in customer config var)

  2. Customer Config Variable (Test Instance)
     - Linked to the scoped config variable template
     - isTest: true (for development/testing)
     - Contains actual credential values (api_key, base_url, etc.)

  This pattern enables:
  - Per-customer connection instances
  - Org-managed credentials (customers don't see/configure them)
  - Test connections for development without a real customer

USAGE IN CNI:
  Reference with: organizationActivatedConnection({ stableKey: "acme-api-key" })

GRAPHQL MUTATIONS:
  1. createScopedConfigVariable (template)
  2. createCustomerConfigVariable (test instance with values)

RETURNS (JSON):
  {
    "stable_key": "acme-api-key",
    "name": "Acme Demo Connection",
    "scoped_config_variable_id": "U2NvcGVkQ29uZmlnVmFyaWFibGU6...",
    "customer_config_variable_id": "Q3VzdG9tZXJDb25maWdWYXJpYWJsZTo...",
    "is_test": true,
    "already_existed": false
  }

EXIT CODES:
  0 - Success
  1 - Authentication error
  3 - API error
"""

import argparse
import json
import os
import sys

SHARED_DIR = os.path.join(os.path.dirname(__file__), '..', 'shared')
sys.path.insert(0, SHARED_DIR)

from graphql import graphql, ensure_authenticated, GraphQLError


def find_connection(component_key, connection_key):
    """Find a connection definition from a component."""
    query = """
    query listComponents($key: String) {
      components(key: $key) {
        nodes {
          id
          key
          connections {
            nodes {
              id
              key
              label
              inputs {
                nodes {
                  key
                  label
                  type
                  default
                }
              }
            }
          }
        }
      }
    }
    """

    try:
        data = graphql(query, {"key": component_key})
    except GraphQLError as e:
        return None, str(e)

    try:
        components = data.get("components", {}).get("nodes", [])
        for component in components:
            if component["key"] == component_key:
                for connection in component["connections"]["nodes"]:
                    if connection["key"] == connection_key:
                        return connection, None
    except (KeyError, TypeError):
        pass

    return None, f"Connection '{connection_key}' not found in component '{component_key}'"


def check_existing_scoped_config_variable(stable_key):
    """Check if a scoped config variable already exists with the given stable key."""
    query = """
    query listScopedConfigVariables($stableKey: String) {
      scopedConfigVariables(stableKey: $stableKey) {
        nodes {
          id
          key
          stableKey
          description
          variableScope
          managedBy
          status
          customerConfigVariables {
            nodes {
              id
              isTest
              status
            }
          }
        }
      }
    }
    """

    try:
        data = graphql(query, {"stableKey": stable_key})
    except GraphQLError:
        return None

    try:
        configs = data.get("scopedConfigVariables", {}).get("nodes", [])
        for config in configs:
            if config["stableKey"] == stable_key:
                return config
    except (KeyError, TypeError):
        pass

    return None


def create_scoped_config_variable(
    connection_def, name, stable_key, description=None
):
    """Create a scoped config variable template (NO values in inputs).

    For variableScope: customer, inputs must NOT have values - only meta.
    The actual credential values are set in the customer config variable.
    """

    # Build inputs with meta and empty values
    # Note: API requires 'value' field even for customer-scoped variables
    inputs = []
    for input_node in connection_def.get("inputs", {}).get("nodes", []):
        input_entry = {
            "name": input_node["key"],
            "type": "value",
            "value": "",  # Empty placeholder - actual values set in customer config var
            "meta": json.dumps({"managedBy": "org", "inputScope": "customer"})
        }
        inputs.append(input_entry)

    mutation = """
    mutation createScopedConfigVariable(
      $key: String!
      $description: String!
      $stableKey: String!
      $variableScope: String!
      $managedBy: String!
      $connection: ID!
      $inputs: [InputExpression]
    ) {
      createScopedConfigVariable(
        input: {
          key: $key
          description: $description
          stableKey: $stableKey
          variableScope: $variableScope
          managedBy: $managedBy
          connection: $connection
          inputs: $inputs
        }
      ) {
        scopedConfigVariable {
          id
          key
          stableKey
          variableScope
          managedBy
          status
        }
        errors {
          field
          messages
        }
      }
    }
    """

    variables = {
        "key": name,
        "description": description or f"Organization connection for {name}",
        "stableKey": stable_key,
        "variableScope": "customer",  # Must be lowercase string
        "managedBy": "org",  # Must be lowercase string
        "connection": connection_def["id"],
        "inputs": inputs,
    }

    try:
        data = graphql(mutation, variables, timeout=60)
    except GraphQLError as e:
        return None, str(e)

    try:
        mutation_result = data.get("createScopedConfigVariable", {})

        if mutation_result.get("errors"):
            error_msgs = []
            for err in mutation_result["errors"]:
                error_msgs.append(f"{err['field']}: {', '.join(err['messages'])}")
            return None, "; ".join(error_msgs)

        return mutation_result["scopedConfigVariable"], None
    except (KeyError, TypeError) as e:
        return None, f"Unexpected response format: {e}"


def find_existing_test_customer_config_variable(scoped_config_var):
    """Find an existing test customer config variable for this scoped var.

    Args:
        scoped_config_var: The scoped config variable dict (from query)

    Returns:
        The customer config variable dict if found, None otherwise
    """
    try:
        customer_vars = scoped_config_var.get("customerConfigVariables", {}).get("nodes", [])
        for var in customer_vars:
            if var.get("isTest") is True:
                return var
    except (KeyError, TypeError):
        pass

    return None


def create_customer_config_variable(
    scoped_config_var_id, field_values=None
):
    """Create a test customer config variable WITH actual credential values.

    This is where the actual API keys and URLs are stored.

    Args:
        scoped_config_var_id: ID of the scoped config variable template
        field_values: Dict of input name -> value (e.g., {"api_key": "...", "app_base_url": "..."})
    """

    # Build inputs WITH actual values
    inputs = []
    if field_values:
        for name, value in field_values.items():
            if value is not None:
                inputs.append({
                    "name": name,
                    "type": "value",
                    "value": value
                })

    mutation = """
    mutation createCustomerConfigVariable(
      $scopedConfigVariable: ID!
      $customer: ID
      $isTest: Boolean
      $inputs: [InputExpression]
    ) {
      createCustomerConfigVariable(
        input: {
          scopedConfigVariable: $scopedConfigVariable
          customer: $customer
          isTest: $isTest
          inputs: $inputs
        }
      ) {
        customerConfigVariable {
          id
          isTest
          status
        }
        errors {
          field
          messages
        }
      }
    }
    """

    variables = {
        "scopedConfigVariable": scoped_config_var_id,
        "customer": None,  # null for test connections
        "isTest": True,
        "inputs": inputs,
    }

    try:
        data = graphql(mutation, variables, timeout=60)
    except GraphQLError as e:
        return None, str(e)

    try:
        mutation_result = data.get("createCustomerConfigVariable", {})

        if mutation_result.get("errors"):
            error_msgs = []
            for err in mutation_result["errors"]:
                error_msgs.append(f"{err['field']}: {', '.join(err['messages'])}")
            return None, "; ".join(error_msgs)

        return mutation_result["customerConfigVariable"], None
    except (KeyError, TypeError) as e:
        return None, f"Unexpected response format: {e}"


def main():
    parser = argparse.ArgumentParser(
        description="Create organization-managed connection via Prismatic GraphQL API"
    )
    parser.add_argument(
        "--component-key",
        required=True,
        help="Component key (e.g., acme, tulyp)",
    )
    parser.add_argument(
        "--connection-key",
        required=True,
        help="Connection key from component (e.g., acmeApiKey)",
    )
    parser.add_argument(
        "--name",
        required=True,
        help="Human-readable name for the connection",
    )
    parser.add_argument(
        "--api-key",
        default=None,
        help="API key value to set",
    )
    parser.add_argument(
        "--base-url",
        default=None,
        help="Base URL value to set",
    )
    parser.add_argument(
        "--stable-key",
        default=None,
        help="Stable key (defaults to {component-key}-api-key)",
    )
    parser.add_argument(
        "--skip-test-connection",
        action="store_true",
        help="Skip creating test customer config variable",
    )

    args = parser.parse_args()

    # Generate stable key if not provided
    stable_key = args.stable_key or f"{args.component_key}-api-key"

    print("=" * 60)
    print("  Create Organization Connection (Scoped)")
    print("=" * 60)
    print("")
    print(f"Component: {args.component_key}")
    print(f"Connection: {args.connection_key}")
    print(f"Name: {args.name}")
    print(f"Stable Key: {stable_key}")
    print("")

    # Verify authentication
    try:
        ensure_authenticated()
    except GraphQLError:
        print("Error: Not authenticated with Prismatic")
        print("Run 'prism login' first")
        return 1

    print("Authenticated with Prismatic")

    # Track what we created vs found
    scoped_var_existed = False
    customer_var_existed = False
    scoped_config_var = None
    customer_config_var = None

    # Check if scoped config variable already exists
    print("Checking for existing scoped config variable...")
    existing_scoped = check_existing_scoped_config_variable(stable_key)

    if existing_scoped:
        print(f"Scoped config variable already exists: {existing_scoped['key']} (ID: {existing_scoped['id']})")
        scoped_config_var = existing_scoped
        scoped_var_existed = True

        # Check for existing test customer config var
        existing_test = find_existing_test_customer_config_variable(existing_scoped)
        if existing_test:
            print(f"Test customer config variable already exists: {existing_test['id']}")
            customer_config_var = existing_test
            customer_var_existed = True
    else:
        # Find the connection definition from the component
        print(f"Finding connection '{args.connection_key}' in component '{args.component_key}'...")
        connection_def, error = find_connection(args.component_key, args.connection_key)

        if error:
            print(f"Error: {error}")
            return 3

        print(f"Found connection: {connection_def['label']}")

        # Create the scoped config variable
        print("Creating scoped config variable...")
        scoped_config_var, error = create_scoped_config_variable(
            connection_def, args.name, stable_key
        )

        if error:
            print(f"Error creating scoped config variable: {error}")
            return 3

        print(f"Created scoped config variable: {scoped_config_var['id']}")

    # Create test customer config variable if not skipped and doesn't exist
    if not args.skip_test_connection and not customer_var_existed:
        # Build field values
        field_values = {}
        if args.api_key:
            field_values["api_key"] = args.api_key
        if args.base_url:
            field_values["app_base_url"] = args.base_url

        print("Creating test customer config variable...")
        customer_config_var, error = create_customer_config_variable(
            scoped_config_var["id"], field_values
        )

        if error:
            print(f"Error creating customer config variable: {error}")
            return 3

        print(f"Created test customer config variable: {customer_config_var['id']}")

    print("")
    print("=" * 60)
    print("  CONNECTION CREATED")
    print("=" * 60)
    print("")

    # Output JSON result
    result = {
        "stable_key": stable_key,
        "name": args.name,
        "scoped_config_variable_id": scoped_config_var["id"],
        "scoped_var_already_existed": scoped_var_existed,
    }

    if customer_config_var:
        result["customer_config_variable_id"] = customer_config_var["id"]
        result["is_test"] = True
        result["customer_var_already_existed"] = customer_var_existed

    print(json.dumps(result, indent=2))

    print("")
    print("Use this stable_key in your CNI integration:")
    print(f'  organizationActivatedConnection({{ stableKey: "{stable_key}" }})')

    return 0


if __name__ == "__main__":
    sys.exit(main())

File: scripts/components/create-organization-connection.ts
#!/usr/bin/env npx tsx
/**
 * create-organization-connection.ts
 *
 * PURPOSE: Create organization-managed connections via Prismatic GraphQL API
 *
 * USAGE: npx tsx create-organization-connection.ts \
 *   --component-key <key> \
 *   --connection-key <key> \
 *   --name <name> \
 *   [--api-key <value>] \
 *   [--base-url <value>] \
 *   [--stable-key <key>] \
 *   [--skip-test-connection]
 *
 * EXIT CODES:
 *   0 - Success
 *   1 - Authentication error
 *   3 - API error
 */

import { graphql, ensureAuthenticated, GraphQLError } from "../shared/graphql.js";

const FIND_CONNECTION_QUERY = `
query listComponents($key: String) {
  components(key: $key) {
    nodes {
      id
      key
      connections {
        nodes {
          id
          key
          label
          inputs {
            nodes {
              key
              label
              type
              default
            }
          }
        }
      }
    }
  }
}
`;

const CHECK_EXISTING_QUERY = `
query listScopedConfigVariables($stableKey: String) {
  scopedConfigVariables(stableKey: $stableKey) {
    nodes {
      id
      key
      stableKey
      description
      variableScope
      managedBy
      status
      customerConfigVariables {
        nodes {
          id
          isTest
          status
        }
      }
    }
  }
}
`;

const CREATE_SCOPED_MUTATION = `
mutation createScopedConfigVariable(
  $key: String!
  $description: String!
  $stableKey: String!
  $variableScope: String!
  $managedBy: String!
  $connection: ID!
  $inputs: [InputExpression]
) {
  createScopedConfigVariable(
    input: {
      key: $key
      description: $description
      stableKey: $stableKey
      variableScope: $variableScope
      managedBy: $managedBy
      connection: $connection
      inputs: $inputs
    }
  ) {
    scopedConfigVariable {
      id
      key
      stableKey
      variableScope
      managedBy
      status
    }
    errors {
      field
      messages
    }
  }
}
`;

const CREATE_CUSTOMER_MUTATION = `
mutation createCustomerConfigVariable(
  $scopedConfigVariable: ID!
  $customer: ID
  $isTest: Boolean
  $inputs: [InputExpression]
) {
  createCustomerConfigVariable(
    input: {
      scopedConfigVariable: $scopedConfigVariable
      customer: $customer
      isTest: $isTest
      inputs: $inputs
    }
  ) {
    customerConfigVariable {
      id
      isTest
      status
    }
    errors {
      field
      messages
    }
  }
}
`;

interface ConnectionDef {
  id: string;
  key: string;
  label: string;
  inputs: { nodes: Array<{ key: string; label: string; type: string; default: unknown }> };
}

interface ScopedConfigVar {
  id: string;
  key: string;
  stableKey: string;
  customerConfigVariables?: { nodes: Array<{ id: string; isTest: boolean; status: string }> };
}

function findConnection(
  componentKey: string,
  connectionKey: string
): { connection: ConnectionDef | null; error: string | null } {
  try {
    const data = graphql(FIND_CONNECTION_QUERY, { key: componentKey }) as Record<string, unknown>;
    const components = ((data.components as Record<string, unknown>)?.nodes ?? []) as Array<Record<string, unknown>>;

    for (const component of components) {
      if (component.key === componentKey) {
        const connections = ((component.connections as Record<string, unknown>)?.nodes ?? []) as ConnectionDef[];
        for (const conn of connections) {
          if (conn.key === connectionKey) {
            return { connection: conn, error: null };
          }
        }
      }
    }
  } catch (e) {
    if (e instanceof GraphQLError) {
      return { connection: null, error: e.message };
    }
  }

  return {
    connection: null,
    error: `Connection '${connectionKey}' not found in component '${componentKey}'`,
  };
}

function checkExistingScopedConfigVariable(stableKey: string): ScopedConfigVar | null {
  try {
    const data = graphql(CHECK_EXISTING_QUERY, { stableKey }) as Record<string, unknown>;
    const configs = ((data.scopedConfigVariables as Record<string, unknown>)?.nodes ?? []) as ScopedConfigVar[];
    for (const config of configs) {
      if (config.stableKey === stableKey) return config;
    }
  } catch {
    // ignore
  }
  return null;
}

function createScopedConfigVariable(
  connectionDef: ConnectionDef,
  name: string,
  stableKey: string,
  description?: string
): { result: Record<string, unknown> | null; error: string | null } {
  const inputs = (connectionDef.inputs?.nodes ?? []).map((inp) => ({
    name: inp.key,
    type: "value",
    value: "",
    meta: JSON.stringify({ managedBy: "org", inputScope: "customer" }),
  }));

  const variables = {
    key: name,
    description: description || `Organization connection for ${name}`,
    stableKey,
    variableScope: "customer",
    managedBy: "org",
    connection: connectionDef.id,
    inputs,
  };

  try {
    const data = graphql(CREATE_SCOPED_MUTATION, variables, 60) as Record<string, unknown>;
    const mutationResult = (data.createScopedConfigVariable ?? {}) as Record<string, unknown>;

    const errors = mutationResult.errors as Array<{ field: string; messages: string[] }> | undefined;
    if (errors?.length) {
      const msgs = errors.map((e) => `${e.field}: ${e.messages.join(", ")}`);
      return { result: null, error: msgs.join("; ") };
    }

    return { result: mutationResult.scopedConfigVariable as Record<string, unknown>, error: null };
  } catch (e) {
    if (e instanceof GraphQLError) return { result: null, error: e.message };
    return { result: null, error: String(e) };
  }
}

function findExistingTestCustomerConfigVariable(
  scopedConfigVar: ScopedConfigVar
): { id: string } | null {
  const customerVars = scopedConfigVar.customerConfigVariables?.nodes ?? [];
  for (const v of customerVars) {
    if (v.isTest === true) return v;
  }
  return null;
}

function createCustomerConfigVariable(
  scopedConfigVarId: string,
  fieldValues?: Record<string, string>
): { result: Record<string, unknown> | null; error: string | null } {
  const inputs: Array<{ name: string; type: string; value: string }> = [];
  if (fieldValues) {
    for (const [name, value] of Object.entries(fieldValues)) {
      if (value != null) {
        inputs.push({ name, type: "value", value });
      }
    }
  }

  const variables = {
    scopedConfigVariable: scopedConfigVarId,
    customer: null,
    isTest: true,
    inputs,
  };

  try {
    const data = graphql(CREATE_CUSTOMER_MUTATION, variables, 60) as Record<string, unknown>;
    const mutationResult = (data.createCustomerConfigVariable ?? {}) as Record<string, unknown>;

    const errors = mutationResult.errors as Array<{ field: string; messages: string[] }> | undefined;
    if (errors?.length) {
      const msgs = errors.map((e) => `${e.field}: ${e.messages.join(", ")}`);
      return { result: null, error: msgs.join("; ") };
    }

    return { result: mutationResult.customerConfigVariable as Record<string, unknown>, error: null };
  } catch (e) {
    if (e instanceof GraphQLError) return { result: null, error: e.message };
    return { result: null, error: String(e) };
  }
}

function parseArgs(args: string[]): {
  componentKey: string;
  connectionKey: string;
  name: string;
  apiKey?: string;
  baseUrl?: string;
  stableKey?: string;
  skipTestConnection: boolean;
} {
  let componentKey = "";
  let connectionKey = "";
  let name = "";
  let apiKey: string | undefined;
  let baseUrl: string | undefined;
  let stableKey: string | undefined;
  let skipTestConnection = false;

  let i = 0;
  while (i < args.length) {
    switch (args[i]) {
      case "--component-key":
        componentKey = args[++i] ?? "";
        break;
      case "--connection-key":
        connectionKey = args[++i] ?? "";
        break;
      case "--name":
        name = args[++i] ?? "";
        break;
      case "--api-key":
        apiKey = args[++i];
        break;
      case "--base-url":
        baseUrl = args[++i];
        break;
      case "--stable-key":
        stableKey = args[++i];
        break;
      case "--skip-test-connection":
        skipTestConnection = true;
        break;
    }
    i++;
  }

  return { componentKey, connectionKey, name, apiKey, baseUrl, stableKey, skipTestConnection };
}

function main(): number {
  const parsed = parseArgs(process.argv.slice(2));

  if (!parsed.componentKey || !parsed.connectionKey || !parsed.name) {
    console.error(
      "Usage: npx tsx create-organization-connection.ts --component-key <key> --connection-key <key> --name <name>"
    );
    return 1;
  }

  const stableKey = parsed.stableKey || `${parsed.componentKey}-api-key`;

  console.log("=".repeat(60));
  console.log("  Create Organization Connection (Scoped)");
  console.log("=".repeat(60));
  console.log("");
  console.log(`Component: ${parsed.componentKey}`);
  console.log(`Connection: ${parsed.connectionKey}`);
  console.log(`Name: ${parsed.name}`);
  console.log(`Stable Key: ${stableKey}`);
  console.log("");

  // Verify authentication
  try {
    ensureAuthenticated();
  } catch {
    console.log("Error: Not authenticated with Prismatic");
    console.log("Run 'prism login' first");
    return 1;
  }

  console.log("Authenticated with Prismatic");

  let scopedVarExisted = false;
  let customerVarExisted = false;
  let scopedConfigVar: Record<string, unknown> | null = null;
  let customerConfigVar: Record<string, unknown> | null = null;

  // Check if scoped config variable already exists
  console.log("Checking for existing scoped config variable...");
  const existingScoped = checkExistingScopedConfigVariable(stableKey);

  if (existingScoped) {
    console.log(
      `Scoped config variable already exists: ${existingScoped.key} (ID: ${existingScoped.id})`
    );
    scopedConfigVar = existingScoped as unknown as Record<string, unknown>;
    scopedVarExisted = true;

    const existingTest = findExistingTestCustomerConfigVariable(existingScoped);
    if (existingTest) {
      console.log(
        `Test customer config variable already exists: ${existingTest.id}`
      );
      customerConfigVar = existingTest as unknown as Record<string, unknown>;
      customerVarExisted = true;
    }
  } else {
    console.log(
      `Finding connection '${parsed.connectionKey}' in component '${parsed.componentKey}'...`
    );
    const { connection, error } = findConnection(
      parsed.componentKey,
      parsed.connectionKey
    );

    if (error || !connection) {
      console.log(`Error: ${error}`);
      return 3;
    }

    console.log(`Found connection: ${connection.label}`);

    console.log("Creating scoped config variable...");
    const createResult = createScopedConfigVariable(
      connection,
      parsed.name,
      stableKey
    );

    if (createResult.error || !createResult.result) {
      console.log(`Error creating scoped config variable: ${createResult.error}`);
      return 3;
    }

    scopedConfigVar = createResult.result;
    console.log(`Created scoped config variable: ${scopedConfigVar.id}`);
  }

  // Create test customer config variable if not skipped and doesn't exist
  if (!parsed.skipTestConnection && !customerVarExisted && scopedConfigVar) {
    const fieldValues: Record<string, string> = {};
    if (parsed.apiKey) fieldValues.api_key = parsed.apiKey;
    if (parsed.baseUrl) fieldValues.app_base_url = parsed.baseUrl;

    console.log("Creating test customer config variable...");
    const { result, error } = createCustomerConfigVariable(
      scopedConfigVar.id as string,
      fieldValues
    );

    if (error || !result) {
      console.log(`Error creating customer config variable: ${error}`);
      return 3;
    }

    customerConfigVar = result;
    console.log(`Created test customer config variable: ${customerConfigVar.id}`);
  }

  console.log("");
  console.log("=".repeat(60));
  console.log("  CONNECTION CREATED");
  console.log("=".repeat(60));
  console.log("");

  const output: Record<string, unknown> = {
    stable_key: stableKey,
    name: parsed.name,
    scoped_config_variable_id: scopedConfigVar?.id,
    scoped_var_already_existed: scopedVarExisted,
  };

  if (customerConfigVar) {
    output.customer_config_variable_id = customerConfigVar.id;
    output.is_test = true;
    output.customer_var_already_existed = customerVarExisted;
  }

  console.log(JSON.stringify(output, null, 2));
  console.log("");
  console.log("Use this stable_key in your CNI integration:");
  console.log(`  organizationActivatedConnection({ stableKey: "${stableKey}" })`);

  return 0;
}

process.exit(main());
